File: src/api/homework.ts
import { RequestFN } from "~/core/request-function";
import { type SessionHandle, type Homework, TabLocation } from "~/models";
import { getUTCDate } from "./private/dates";
import { translateToWeekNumber } from "./helpers/week-number";
import { decodeHomework } from "~/decoders/homework";

export const homeworkFromWeek = async (session: SessionHandle, weekNumber: number, extendsToWeekNumber?: number): Promise<Array<Homework>> => {
  const request = new RequestFN(session, "PageCahierDeTexte", {
    _Signature_: { onglet: TabLocation.Homework },
    donnees: {
      domaine: {
        _T: 8,
        V: typeof extendsToWeekNumber === "number" ? `[${weekNumber}..${extendsToWeekNumber}]` : `[${weekNumber}]`
      }
    }
  });

  const response = await request.send();
  return response.data.donnees.ListeTravauxAFaire.V.map(decodeHomework);
};

export const homeworkFromIntervals = async (session: SessionHandle, startDate: Date, endDate: Date): Promise<Array<Homework>> => {
  startDate = getUTCDate(startDate);
  endDate = getUTCDate(endDate);

  const startWeekNumber = translateToWeekNumber(startDate, session.instance.firstMonday);
  const endWeekNumber = translateToWeekNumber(endDate, session.instance.firstMonday);

  const reply = await homeworkFromWeek(session, startWeekNumber, endWeekNumber);

  // Only keep items homework are in the intervals.
  return reply.filter((homework) => startDate <= homework.deadline && homework.deadline <= endDate);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/private/homework.kt
package ink.literate.pawnote.api.private

import ink.literate.pawnote.api.helpers.translateToWeekNumber
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.SessionInformation
import ink.literate.pawnote.models.TabLocation

import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun homeworkFromWeek (sessionInformation: SessionInformation, tab: TabLocation, weekNumber: Int, extendsToWeekNumber: Int? = null): JsonObject {
    val request = RequestFN(sessionInformation, "PageCahierDeTexte", Json.encodeToString(buildJsonObject {
        putJsonObject("_Signature_") {
            put("onglet", tab.code)
        }

        putJsonObject("donnees") {
            putJsonObject("domaine") {
                put("_T", 8)
                put("V", if (extendsToWeekNumber != null) "[$weekNumber..$extendsToWeekNumber]" else "[$weekNumber]")
            }
        }
    }))

    val response = request.send()
    return Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject
}

suspend fun homeworkFromIntervals (session: SessionHandle, tab: TabLocation, startDate: LocalDateTime, endDate: LocalDateTime): JsonObject {
    val startWeekNumber = translateToWeekNumber(startDate, session.instance.firstMonday)
    val endWeekNumber = translateToWeekNumber(endDate, session.instance.firstMonday)

    return homeworkFromWeek(session.information, tab, startWeekNumber, endWeekNumber)
}